net: lib: sntp: async query API

JordanYates opened the pull request:
Add a variant of the SNTP API that does not block while waiting for a
response from the SNTP server. Instead it takes advantage of
`CONFIG_NET_SOCKETS_SERVICE` to read the response asynchronously when
the response is received.

Example output:
```
> west build -b native_sim -t run zephyr/samples/net/sntp/ -T net.sntp.native -p
...
*** Booting Zephyr OS build v4.0.0-5533-g499ba806a4e9 ***
[00:00:00.000,000] <inf> net_config: Initializing network
[00:00:00.000,000] <inf> net_config: IPv4 address: 192.0.2.1
[00:00:00.000,000] <inf> net_sntp_sample: Starting SNTP sample
[00:00:00.000,000] <inf> net_samples_common: Network connectivity established and IP address assigned
[00:00:00.000,000] <inf> net_samples_common: Waiting for network...
[00:00:00.000,000] <inf> net_sntp_sample: pool.ntp.org -> 129.250.35.251:123
[00:00:00.000,000] <inf> net_sntp_sample: SNTP query using synchronous APIs
[00:00:00.180,003] <inf> net_sntp_sample:  SNTP Time: 1739709549
[00:00:00.180,003] <inf> net_sntp_sample: Local Time: 1739709549
[00:00:01.190,000] <inf> net_sntp_sample: SNTP query using asynchronous APIs
[00:00:01.350,003] <inf> net_sntp_sample:  SNTP Time: 1739709550
[00:00:01.350,003] <inf> net_sntp_sample: Local Time: 1739709550
[00:00:01.350,003] <inf> net_sntp_sample: SNTP sample complete
```

Recreation of #79445 with the tests converted to a sample.

JordanYates committed:
net: lib: sntp: extract query send logic

Extract the SNTP query send logic out of `sntp_query` so that it can be
used by functions that don't synchronously wait for the response.

Signed-off-by: Jordan Yates <jordan@embeint.com>

JordanYates committed:
net: lib: sntp: async query API

Add a variant of the SNTP API that does not block while waiting for a
response from the SNTP server. Instead it takes advantage of
`CONFIG_NET_SOCKETS_SERVICE` to read the response asynchronously when
the response is received.

Signed-off-by: Jordan Yates <jordan@embeint.com>

JordanYates committed:
samples: net: common: resend conn status

Networking events may have been triggered before `wait_for_network` was
called. Resend the status events after the callbacks have been
registered.

Signed-off-by: Jordan Yates <jordan@embeint.com>

jukkar commented on samples/net/sntp/src/main.c:
```diff
@@ -0,0 +1,174 @@
+	/* Free the allocated memory */
+	zsock_freeaddrinfo(res);
+
+	if (addr->sa_family == AF_INET) {
+		struct sockaddr_in *ipv4 = (struct sockaddr_in *)addr;
+		uint8_t *b = ipv4->sin_addr.s4_addr;
+
+		ipv4->sin_port = htons(port);
+		LOG_INF("%s -> %d.%d.%d.%d:%d", host, b[0], b[1], b[2], b[3], port);
+	} else {
+		struct sockaddr_in6 *ipv6 = (struct sockaddr_in6 *)addr;
+
+		ipv6->sin6_port = htons(port);
+		LOG_INF("%s -> IPv6:%d", host, port);
+	}
```
This looks a bit of a kludge, I mean there is `inet_ntop` that can create the printable ip address, and we don't even need this if()else thingy.

JordanYates replied:
That function exists in `<zephyr/posix/arpa/inet.h>`, is there any expectation that the function will exist on non-posix targets?

jukkar commented on samples/net/sntp/src/main.c:
```diff
@@ -0,0 +1,174 @@
+int dns_query(const char *host, uint16_t port, int family, int socktype, struct sockaddr *addr,
+	      socklen_t *addrlen)
+{
+	struct zsock_addrinfo hints = {
```
For the networking samples, we have tried to mainly use the BSD socket API (so calling without `zsock_` which means that `CONFIG_POSIX_API` needs to be enabled).

JordanYates replied:
The downside for this sample is that `CONFIG_POSIX_API` looks to break the `time` function, meaning you can no longer compare the SNTP time to the local system time if run on `native_posix`.

jukkar replied:
Ok, fair enough, there is certainly things in Posix that do not work properly yet.

JordanYates commented:
Updated IP address printing to use `inet_ntop`

jukkar commented on samples/net/sntp/src/main.c:
```diff
@@ -0,0 +1,177 @@
+	/* Free the allocated memory */
+	zsock_freeaddrinfo(res);
+
+	if (addr->sa_family == AF_INET) {
+		struct sockaddr_in *ipv4 = (struct sockaddr_in *)addr;
+
+		ipv4->sin_port = htons(port);
+		inet_ntop(ipv4->sin_family, &ipv4->sin_addr, addr_str, sizeof(addr_str));
+	} else {
+		struct sockaddr_in6 *ipv6 = (struct sockaddr_in6 *)addr;
+
+		ipv6->sin6_port = htons(port);
+		inet_ntop(ipv6->sin6_family, &ipv6->sin6_addr, addr_str, sizeof(addr_str));
+	}
```
I think we could replace this by so no need to check the family here.
```
net_sin(addr)->sin_port = htons(port);
zsock_inet_ntop(addr->sa_family, &net_sin(addr)->sin_addr, addr_str, sizeof(addr_str));
```
This works because the port number and address are pointing to the same place both in `sockaddr_in` and `sockaddr_in6`
Also there is a zsock variant for inet_ntop so that you do not need to include the `arpa/inet.h`

rlubos reviewed (commented):
Looks good, regarding the sample however, wouldn't it make sense to extend https://github.com/zephyrproject-rtos/zephyr/tree/main/samples/net/sockets/sntp_client instead?

JordanYates committed:
samples: net: sntp_client: native sockets

Add the option to use the native socket abstraction for the sample.

Signed-off-by: Jordan Yates <jordan@embeint.com>

JordanYates committed:
samples: net: sntp_client: SNTP server lookup

Lookup the address of the SNTP server through DNS instead of requiring
a server to be running locally.

Signed-off-by: Jordan Yates <jordan@embeint.com>

JordanYates committed:
samples: net: sntp_client: demonstrate async SNTP

Demonstrate asynchronous SNTP querying in the sample.

Signed-off-by: Jordan Yates <jordan@embeint.com>

JordanYates commented:
> Looks good, regarding the sample however, wouldn't it make sense to extend https://github.com/zephyrproject-rtos/zephyr/tree/main/samples/net/sockets/sntp_client instead?

Good point, I think I burned that sample from my memory years ago when I saw that it required running a local SNTP server :).
